File: app/src/main/java/com/omarea/vboot/reciver_batterychanged.java
package com.omarea.vboot;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.BatteryManager;
import android.os.Handler;
import android.os.Message;
import android.widget.Toast;

import com.omarea.shared.AppShared;
import com.omarea.shared.ConfigInfo;
import com.omarea.shared.Consts;
import com.omarea.shared.EventBus;
import com.omarea.shared.Events;
import com.omarea.shared.ServiceHelper;

import java.io.DataOutputStream;
import java.io.IOException;

/**
 * Created by helloklf on 2016/12/21.
 */


public class reciver_batterychanged extends BroadcastReceiver {
    public static ServiceHelper serviceHelper;

    private void DoCmd(String cmd) {
        Process p = null;
        try {
            try {
                p = Runtime.getRuntime().exec("su");
                DataOutputStream out = new DataOutputStream(p.getOutputStream());
                out.writeBytes(cmd);
                out.writeBytes("\n");
                out.writeBytes("exit\n");
                out.writeBytes("exit\n");
                out.flush();
                p.waitFor();
            } catch (IOException e) {
                if (p != null) {
                    p.destroy();
                    p = null;
                }
                NoRoot();
            } catch (Exception ex) {
                if (p != null) {
                    p.destroy();
                    p = null;
                }
                ShowMsg("电量监控服务异常：" + ex.getMessage(), false);
            } finally {
                if (p != null)
                    p.destroy();
            }
        } catch (Exception ex) {
            ShowMsg("电量监控服务异常：" + ex.getMessage(), false);
        }
    }

    Context context;
    private Handler myHandler = new Handler() {
        public void handleMessage(Message msg) {
            super.handleMessage(msg);
        }
    };

    //显示文本消息
    private void ShowMsg(final String msg, final boolean longMsg) {
        if (context != null)
            myHandler.post(new Runnable() {
                @Override
                public void run() {
                    Toast.makeText(context, msg, longMsg ? Toast.LENGTH_LONG : Toast.LENGTH_SHORT).show();
                }
            });
    }

    private void NoRoot() {
        ShowMsg("充电加速服务异常，请检查是否已ROOT手机，并允许本应用访问ROOT权限！", false);
    }

    //快速充电
    private void FastCharger() {
        if (!ConfigInfo.getConfigInfo().QcMode)
            return;
        DoCmd(Consts.FastChanger);
    }

    /**
     * 电池保护
     */
    private void BP() {
        DoCmd(Consts.BP);
    }

    private void BPReset() {
        DoCmd(Consts.BPReset);
    }

    int lastBatteryLeavel = -1;
    boolean lastChangerState = false;

    @Override
    public void onReceive(Context context, Intent intent) {
        this.context = context;
        try {
            String action = intent.getAction();
            boolean onChanger = intent.getIntExtra(BatteryManager.EXTRA_STATUS, -1) == BatteryManager.BATTERY_STATUS_CHARGING;
            int batteryLevel = intent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1);//intent.getIntExtra(BatteryManager.EXTRA_SCALE,-1);

            if (lastBatteryLeavel != batteryLevel) {
                EventBus.publish(Events.BatteryChanged, batteryLevel);
                lastBatteryLeavel = batteryLevel;
            }

            if (onChanger && (batteryLevel < 85 || !ConfigInfo.getConfigInfo().BatteryProtection))
                BPReset();

            if (action.equals(Intent.ACTION_BATTERY_CHANGED)) {
                if (AppShared.system_inited && onChanger) {
                    if (ConfigInfo.getConfigInfo().QcMode) FastCharger();
                    if (ConfigInfo.getConfigInfo().BatteryProtection) BP();
                }
            } else if (action.equals(Intent.ACTION_POWER_CONNECTED)) {
            } else if ((action.equals(Intent.ACTION_BOOT_COMPLETED) && onChanger)) {
            } else if (action.equals(Intent.ACTION_POWER_DISCONNECTED)) {
                BPReset();
                EventBus.publish(Events.PowerDisConnection);
                return;
            }

            //如果充电状态切换
            if (lastChangerState != onChanger) {
                lastChangerState = onChanger;
                if (onChanger) {
                    BPReset();
                    entryFastChanger(onChanger);
                    EventBus.publish(Events.PowerConnection);
                } else
                    EventBus.publish(Events.PowerDisConnection);
            }
        } catch (Exception ex) {
        }
    }

    private void entryFastChanger(boolean onChanger) {
        if (AppShared.system_inited && onChanger) {
            if (ConfigInfo.getConfigInfo().QcMode)
                FastCharger();
            if (ConfigInfo.getConfigInfo().BatteryProtection)
                BP();
            if (ConfigInfo.getConfigInfo().DebugMode)
                ShowMsg("充电器已连接！", false);
        }
    }
}

File: app/src/main/java/com/omarea/vboot/ReciverBatterychanged.kt
package com.omarea.vboot

import android.app.Service
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.os.BatteryManager
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.widget.Toast
import com.omarea.shared.Consts
import com.omarea.shared.SpfConfig
import com.omarea.shared.helper.KeepShell
import android.os.AsyncTask



class ReciverBatterychanged(private var service: Service) : BroadcastReceiver() {
    private var bp: Boolean = false
    private var keepShell: KeepShell? = null

    private var sharedPreferences: SharedPreferences
    private var globalSharedPreferences: SharedPreferences
    private var listener: SharedPreferences.OnSharedPreferenceChangeListener
    private val myHandler = Handler(Looper.getMainLooper())
    private var qcLimit = 50000

    //显示文本消息
    private fun showMsg(msg: String, longMsg: Boolean) {
        try {
            myHandler.post {
                Toast.makeText(service, msg, if (longMsg) Toast.LENGTH_LONG else Toast.LENGTH_SHORT).show()
            }
        } catch (ex: Exception) {
            Log.e("BatteryService", ex.message)
        }
    }

    //快速充电
    private fun fastCharger() {
        try {
            if (!sharedPreferences.getBoolean(SpfConfig.CHARGE_SPF_QC_BOOSTER, false))
                return

            if (globalSharedPreferences.getBoolean(SpfConfig.GLOBAL_SPF_DEBUG, false))
                showMsg(service.getString(R.string.power_connected), false)
            keepShell!!.doCmd(Consts.FastChangerBase)
            keepShell!!.doCmd(computeLeves(qcLimit).toString())
        } catch (ex: Exception) {
            Log.e("ChargeService", ex.stackTrace.toString())
        }
    }

    private fun computeLeves(qcLimit: Int): StringBuilder {
        val arr = StringBuilder()
        if (qcLimit > 300) {
            var level = 300
            while (level < qcLimit) {
                arr.append("echo ${level}000 > /sys/class/power_supply/battery/constant_charge_current_max\n")
                arr.append("echo ${level}000 > /sys/class/power_supply/main/constant_charge_current_max\n")
                arr.append("echo ${level}000 > /sys/class/qcom-battery/restricted_current\n")
                level += 300
            }
        }
        arr.append("echo ${qcLimit}000 > /sys/class/power_supply/battery/constant_charge_current_max\n")
        arr.append("echo ${qcLimit}000 > /sys/class/power_supply/main/constant_charge_current_max\n")
        arr.append("echo ${qcLimit}000 > /sys/class/qcom-battery/restricted_current\n")
        return arr;
    }

    override fun onReceive(context: Context, intent: Intent) {
        val pendingResult = goAsync()
        val asyncTask = TaskAsync()
        asyncTask.execute(Runnable {
            onReceiveAsync(intent, pendingResult)
        })
    }

    private fun onReceiveAsync(intent: Intent, pendingResult: PendingResult) {
        try {
            val action = intent.action
            val onChanger = intent.getIntExtra(BatteryManager.EXTRA_STATUS, -1) == BatteryManager.BATTERY_STATUS_CHARGING
            val batteryLevel = intent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1)

            //BatteryProtection
            if (sharedPreferences.getBoolean(SpfConfig.CHARGE_SPF_BP, false)) {
                if (onChanger) {
                    if (batteryLevel >= sharedPreferences.getInt(SpfConfig.CHARGE_SPF_BP_LEVEL, 85)) {
                        bp = true
                        keepShell!!.doCmd(Consts.DisableChanger)
                    } else if (batteryLevel < sharedPreferences.getInt(SpfConfig.CHARGE_SPF_BP_LEVEL, 85) - 20) {
                        resumeCharge()
                    }
                }
                //电量不足，恢复充电功能
                else if (action == Intent.ACTION_BATTERY_LOW) {
                    showMsg(service.getString(R.string.battery_low), false)
                    resumeCharge()
                } else if (bp && batteryLevel != -1 && batteryLevel < sharedPreferences.getInt(SpfConfig.CHARGE_SPF_BP_LEVEL, 85) - 20) {
                    //电量低于保护级别20
                    resumeCharge()
                }
            }

            entryFastChanger(onChanger)
        } catch (ex: Exception) {
            showMsg("充电加速服务：\n" + ex.message, true);
        } finally {
            pendingResult.finish()
        }
    }


    internal class TaskAsync : AsyncTask<Runnable, Int, Unit>() {
        override fun doInBackground(vararg params: Runnable) {
            params[0].run()
        }
    }

    init {
        if (keepShell == null) {
            keepShell = KeepShell(service)
        }
        sharedPreferences = service.getSharedPreferences(SpfConfig.CHARGE_SPF, Context.MODE_PRIVATE)
        listener = SharedPreferences.OnSharedPreferenceChangeListener { spf, key ->
            if (key == SpfConfig.CHARGE_SPF_QC_LIMIT) {
                qcLimit = spf.getInt(SpfConfig.CHARGE_SPF_QC_LIMIT, qcLimit)
            }
        }
        sharedPreferences.registerOnSharedPreferenceChangeListener(listener)
        qcLimit = sharedPreferences.getInt(SpfConfig.CHARGE_SPF_QC_LIMIT, qcLimit)
        globalSharedPreferences = service.getSharedPreferences(SpfConfig.GLOBAL_SPF, Context.MODE_PRIVATE)
    }

    internal fun onDestroy() {
        sharedPreferences.unregisterOnSharedPreferenceChangeListener(this.listener)
        this.resumeCharge()
        keepShell!!.tryExit()
        keepShell = null
    }

    internal fun resumeCharge() {
        bp = false
        keepShell!!.doCmd(Consts.ResumeChanger)
    }

    internal fun entryFastChanger(onChanger: Boolean) {
        if (onChanger) {
            if (sharedPreferences.getBoolean(SpfConfig.CHARGE_SPF_QC_BOOSTER, false)) {
                fastCharger()
            }
        }
    }
}
